In [17]:
import numpy as np
from sympy import Matrix
import matplotlib.pyplot as plt
import seaborn as sns

In [18]:
ID = "0432310005101041"
d1 = int(ID[-1])
d2 = int(ID[-2])

A = np.array([
    [d1+2 , d2+1],
    [2*d1, d2+2]
])
# Matrix A
print("matrix A: \n", A)

# Analysis of Matrix A
size_A = A.shape
det_A = np.linalg.det(A)
rank_A = np.linalg.matrix_rank(A)
eigenvalues_A, eigenvactors_A = np.linalg.eig(A)

print("\nsize_A:\n", size_A)
print("\ndet_A:\n", det_A)
print("\nrank_A:\n", rank_A)
print("\neigenvalues_A:\n", eigenvalues_A)
print("\neigenvactors_A:\n", eigenvactors_A)

if det_A != 0:
    inv_A = np.linalg.inv(A)
    print("Inverse of A:\n", inv_A)
else:
    print("Inverse not possible")

matrix A: 
 [[3 5]
 [2 6]]

size_A:
 (2, 2)

det_A:
 8.000000000000002

rank_A:
 2

eigenvalues_A:
 [1. 8.]

eigenvactors_A:
 [[-0.92847669 -0.70710678]
 [ 0.37139068 -0.70710678]]
Inverse of A:
 [[ 0.75  -0.625]
 [-0.25   0.375]]


In [19]:
# Value Change Experiment
# Change exactly one element (+1)

B = A.copy()
B[1, 0] += 1   # Changed element A[1, 0] by +1]

print("\nMatrix B after adding +1 into A[1, 0]):\n", B)

# Analysis of Matrix B
size_B = B.shape
det_B = np.linalg.det(B)
rank_B = np.linalg.matrix_rank(B)
eigenvalues_B, eigenvactors_B = np.linalg.eig(B)

print("\nsize_B:\n", size_B)
print("\ndet_B:\n", det_B)
print("\nrank_B:\n", rank_B)
print("\neigenvalues_B:\n", eigenvalues_B)
print("\neigenvactors_B:\n", eigenvactors_B)

if det_B != 0:
    inv_B = np.linalg.inv(B)
    print("Inverse of B:\n", inv_B)
else:
    print("Inverse not possible")



Matrix B after adding +1 into A[1, 0]):
 [[3 5]
 [3 6]]

size_B:
 (2, 2)

det_B:
 3.0000000000000004

rank_B:
 2

eigenvalues_B:
 [0.34668807 8.65331193]

eigenvactors_B:
 [[-0.88333068 -0.66249905]
 [ 0.46875037 -0.74906275]]
Inverse of B:
 [[ 2.         -1.66666667]
 [-1.          1.        ]]


1.How did the determinant change and why?

    The determinant decreased from 8 to 3 after changing one matrix value.
    This happened because modifying a single element changed how the rows interact with each other.

2.Did the rank change? Explain.

    No, the rank did not change and remained 2 for both matrices.
    This is because the rows are still linearly independent after the value change.

3.How did the eigenvalues respond to the value change?

    The eigenvalues changed from [1, 8] to approximately [0.35, 8.65].
    This shows that eigenvalues are sensitive to even small changes in matrix elements.

4.Is B easier or harder to invert than A? Why?

    Matrix B is harder to invert because its determinant is smaller.
    A smaller determinant makes the inverse less stable compared to matrix A.